src/GitVersion.Core/Configuration/ConfigurationHelper.cs

```diff
@@ -0,0 +1,103 @@
+using GitVersion.Extensions;
+
+namespace GitVersion.Configuration;
+
+internal class ConfigurationHelper
+{
+    public string Yaml
+    {
+        get
+        {
+            if (this.yaml == null)
+            {
+                if (this.dictionary == null)
+                {
+                    this.yaml = ConfigurationSerializer.Serialize(this.configuration!);
+                }
+                else
+                {
+                    this.yaml = ConfigurationSerializer.Serialize(this.dictionary);
+                }
+            }
+            return this.yaml;
+        }
+    }
+    private string? yaml;
+
+    public IDictionary<object, object?> Dictionary
+    {
+        get
+        {
+            if (this.dictionary == null)
+            {
+                this.yaml ??= ConfigurationSerializer.Serialize(this.configuration!);
+                this.dictionary = ConfigurationSerializer.Deserialize<Dictionary<object, object?>>(this.yaml!);
+            }
+            return this.dictionary;
+        }
+    }
+    private IDictionary<object, object?>? dictionary;
+
+    public GitVersionConfiguration Configuration => this.configuration ??= ConfigurationSerializer.Deserialize<GitVersionConfiguration>(Yaml);
+    private GitVersionConfiguration? configuration;
+
+    public ConfigurationHelper(string yaml) => this.yaml = yaml.NotNull();
+
+    public ConfigurationHelper(IDictionary<object, object?> dictionary) => this.dictionary = dictionary.NotNull();
+
+    public ConfigurationHelper(GitVersionConfiguration configuration) => this.configuration = configuration.NotNull();
+
+    public void Override(IDictionary<object, object?> dictionary)
+    {
+        dictionary.NotNull();
+
+        if (dictionary.Any())
+        {
+            this.dictionary = Merge(Dictionary, dictionary);
+            this.yaml = null;
+            this.configuration = null;
+        }
+    }
+
+    private static IDictionary<object, object?> Merge(IDictionary<object, object?> dictionary, IDictionary<object, object?> anotherDictionary)
```

Comment:
The Merge happens only on two dictionaries. That's mean we are not merging on GitVersionConfiguration class layer. I thought it's good to place it here but I can also place it some where else. Creating an extensions or static class might be an idea. Please let me know what do you like.

Reply:
I'm also not happy with the generic IDictionary type. The alternative is JObject from Newtonsoft. The usage of the GitVersionConfiguration class is not possible because I cannot make a distinguish between null and not existing.

If you like I can change the type from IDictionary to IReadOnlyDictionary. What do you think?

Reply:
I have changed the override configuration type to IReadOnlyDictionary

Reply:
If the dictionary is there to facilitate deserialization, it's not a core domain concern and should be moved to the perimeter of the application.

Reply:
Of course but this has nothing to do with the PR.

Reply:
Okay, agreed.